pkg/operator/ceph/csi/mapping_config.go

```diff
@@ -0,0 +1,258 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"strconv"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+var (
+	// configuration for mapping ClusterID and RDB PoolID
+	mappingConfigName = "rook-ceph-csi-mapping-config"
+	mappingConfigkey  = "csi-mapping-config-json"
+	cfg               = `
+[global]
+fsid = %s
+mon host = %s
+
+[%s]
+keyring = %s
+`
+)
+
+type CSIDRMapping struct {
+	ClusterIDMapping map[string]string
+	RBDPoolIDMapping []map[string]string
+	//TODO: CephFSFsIDMapping []map[string]string
+}
+
+type CSIDRMappings []CSIDRMapping
+
+func (m *CSIDRMappings) addClusterIDMapping(newClusterIDMap map[string]string) {
+	if !m.hasClusterIDMapping(newClusterIDMap) {
+		newDRMap := CSIDRMapping{
+			ClusterIDMapping: newClusterIDMap,
+		}
+		*m = append(*m, newDRMap)
+	}
+}
+
+func (m *CSIDRMappings) addRBDPoolIDMapping(clusterIDMap, newPoolIDMap map[string]string) {
+	for i := 0; i < len(*m); i++ {
+		if reflect.DeepEqual((*m)[i].ClusterIDMapping, clusterIDMap) {
+			(*m)[i].RBDPoolIDMapping = append((*m)[i].RBDPoolIDMapping, newPoolIDMap)
+		}
+	}
+}
+
+func (m *CSIDRMappings) hasClusterIDMapping(newClusterIDMap map[string]string) bool {
+	for _, mapping := range *m {
+		if reflect.DeepEqual(mapping.ClusterIDMapping, newClusterIDMap) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func ReconcileCSIMapConfig(clusterInfo *cephclient.ClusterInfo, ctx *clusterd.Context, pool *cephv1.CephBlockPool, ownerInfo *k8sutil.OwnerInfo) error {
```

Comment:
Seb started working on a controller that would manage the CSI driver. This may fit into that controller where we generate the CSI configmap whenever the controller sees relevant changes. Let's discuss when he returns.

Reply:
I had already started working on the separate controller for this config map before reading your message. Just pushed those changes yesterday. We can discuss more and maybe move it to the controller that Seb is working on. Shouldn't be a lot of changes.